### PR TITLE
rgw: fix realm reload cause crash in stop rgw

### PR DESCRIPTION
**Pull request by tianshan**
Signed-off-by: Tianshan Qu <tianshan@xsky.com>

### CODE REVIEW

**Comment by tianshan** [comment]
@cbodley  please help review

**Comment by cbodley** [comment]
could you please share a backtrace of the crash?

**Comment by tianshan** [comment]
reproduce this with radosgw-admin period update --commit && killall radosgw
backtrace may different in multi-run. reload and stop rgw will run close_storage both, some structure may be destroyed.

**log:**
2017-12-04 22:20:31.596 7f9abca36700  1 rgw realm reloader: Pausing frontends for realm update...

2017-12-04 22:20:31.670 7f9aefa9c700 -1 received  signal: Terminated from  PID: 69692 task name: killall radosgw  UID: 0

existing dependency complete_op_data (170) -> RGWIndexCompletionManager::lock::61 (124) at:
2017-12-04 22:20:32.212 7f9abca36700  0 new dependency RGWIndexCompletionManager::lock::61 (124) -> complete_op_data (170) creates a cycle at
2017-12-04 22:20:32.212 7f9abca36700  0 btw, i am holding these locks:
2017-12-04 22:20:32.212 7f9abca36700  0   RGWIndexCompletionManager::lock::61 (124)
2017-12-04 22:20:32.212 7f9abca36700  0   RGWCivetWebFrontend (168)
2017-12-04 22:20:32.212 7f9abca36700  0
2017-12-04 22:20:32.214 7f9abca36700 -1 *** Caught signal (Aborted) **

**backtrace**
#0  0x00007f9b21756fcb in raise (sig=6) at ../nptl/sysdeps/unix/sysv/linux/pt-raise.c:37
#1  0x00007f9b22e4a092 in reraise_fatal (signum=6) at /root/ceph-int/src/global/signal_handler.cc:74
#2  handle_fatal_signal	(signum=6) at /root/ceph-int/src/global/signal_handler.cc:138
#3  <signal handler called>
#4  0x00007f9b164ed5f7 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#5  0x00007f9b164eece8 in __GI_abort () at abort.c:90
#6  0x00007f9b190617ad in lockdep_will_lock (name=<optimized out>, id=<optimized out>, force_backtrace=false) at /root/ceph-int/src/common/lockdep.cc:314
#7  0x00007f9b18db4d52 in _will_lock (this=0x7f9b277359c0) at /root/ceph-int/src/common/Mutex.h:56
#8  Mutex::Lock	(this=this@entry=0x7f9b277359c0, no_lockdep=no_lockdep@entry=false) at /root/ceph-int/src/common/Mutex.cc:92
#9  0x00007f9b22f7be56 in Locker (m=..., this=<synthetic pointer>) at /root/ceph-int/src/common/Mutex.h:115
#10 stop (this=0x7f9b25f14e60) at /root/ceph-int/src/rgw/rgw_rados.cc:3549
#11 ~RGWIndexCompletionManager (this=0x7f9b25f14e60, __in_chrg=<optimized out>) at /root/ceph-int/src/rgw/rgw_rados.cc:3507
#12 RGWRados::finalize (this=this@entry=0x7f9b25ce2000) at /root/ceph-int/src/rgw/rgw_rados.cc:3740
#13 0x00007f9b22f7c11e in RGWStoreManager::close_storage (store=0x7f9b25ce2000) at /root/ceph-int/src/rgw/rgw_rados.cc:13826
#14 0x00007f9b22ecf540 in RGWRealmReloader::reload (this=0x7fffc2b0ee20) at /root/ceph-int/src/rgw/rgw_realm_reloader.cc:90
#15 0x00007f9b22df3579 in Context::complete (this=0x7f9b275ece60, r=<optimized out>) at	/root/ceph-int/src/include/Context.h:70
#16 0x00007f9b18ddb020 in SafeTimer::timer_thread (this=0x7fffc2b0ee40) at /root/ceph-int/src/common/Timer.cc:97
#17 0x00007f9b18ddc2ed in SafeTimerThread::entry (this=<optimized out>) at /root/ceph-int/src/common/Timer.cc:30
#18 0x00007f9b2174fdc5 in start_thread (arg=0x7f9abca36700) at pthread_create.c:308
#19 0x00007f9b165ae21d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:113

File: src/rgw/rgw_main.cc [hunk 1]
```diff
@@ -537,6 +537,8 @@ int main(int argc, const char **argv)
 
   derr << "shutting down" << dendl;
 
+  delete reloader;
+
```
Comment (cbodley) [inline src/rgw/rgw_main.cc]: can you help me understand how deleting the reloader early addresses the segfault? also note that `realm_watcher` will still hold a reference to this `reloader`

Comment (tianshan) [reply src/rgw/rgw_main.cc]: Oh, I only tested restart after reload. I think add "realm_watcher.watch_stop()" will fix reload after restart situation, and local test passed. whats your opinion?

**Pull request closed without merging**